File: Ahorn/entities/MagneticCeiling.jl
module XaphanHelperMagneticCeiling

using ..Ahorn, Maple

@mapdef Entity "XaphanHelper/MagneticCeiling" MagneticCeiling(x::Integer, y::Integer, width::Integer=8, directory::String="objects/XaphanHelper/MagneticCeiling", animationSpeed::Number=0.20, canJump::Bool=false, noStaminaDrain::Bool=false)

const placements = Ahorn.PlacementDict(
   "Magnetic Ceiling (Xaphan Helper)" => Ahorn.EntityPlacement(
	MagneticCeiling,
		"rectangle",
		Dict{String, Any}()
	),
)

Ahorn.minimumSize(entity::MagneticCeiling) = 8, 8
Ahorn.resizable(entity::MagneticCeiling) =  true, false

function Ahorn.selection(entity::MagneticCeiling)
    x, y = Ahorn.position(entity)

    width = Int(get(entity.data, "width", 8))

    return return Ahorn.Rectangle(x, y, width, 8)
end

function renderCeiling(ctx::Ahorn.Cairo.CairoContext, x::Number, y::Number, width::Number, directory::String)
    frameA = "$(directory)/idle_a00"
    frameB = "$(directory)/idle_b00"
    
    tilesWidth = div(width, 8)

    for i in 1:tilesWidth
        if iseven(i)
            Ahorn.drawImage(ctx, frameB, x + (i - 1) * 8, y, 0, 0, 8, 8)
        else
            Ahorn.drawImage(ctx, frameA, x + (i - 1) * 8, y, 0, 0, 8, 8)
        end
    end
end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::MagneticCeiling, room::Maple.Room)
    directory = get(entity.data, "directory", "objects/XaphanHelper/MagneticCeiling")
    startX, startY = Int(entity.data["x"]), Int(entity.data["y"])

    width = Int(get(entity.data, "width", 8))

    renderCeiling(ctx, startX, startY, width, directory)
end

end

File: Loenn/entities/MagneticCeiling.lua
local drawableSprite = require("structs.drawable_sprite")

local MagneticCeiling = {}

MagneticCeiling.name = "XaphanHelper/MagneticCeiling"
MagneticCeiling.depth = -9999
MagneticCeiling.fieldOrder = {"x", "y", "width", "directory", "animationSpeed", "canJump", "noStaminaDrain"}
MagneticCeiling.canResize = {true, false}
MagneticCeiling.placements = {
    name = "MagneticCeiling",
    data = {
        width = 8,
        directory = "objects/XaphanHelper/MagneticCeiling",
        animationSpeed = 0.20,
        canJump = false,
        noStaminaDrain = false
    }
}

function MagneticCeiling.sprite(room, entity)
    local sprites = {}

    local directory = entity.directory or "objects/XaphanHelper/MagneticCeiling"
    local width = entity.width / 8 or 1

    local sprite = nil
    for i = 1,width do
        if (i % 2 == 0) then
            sprite = drawableSprite.fromTexture(directory .. "/idle_b00", entity)
        else
            sprite = drawableSprite.fromTexture(directory .. "/idle_a00", entity)
        end
        sprite:addPosition((i - 1) * 8 + 4, 4)
        if sprite then
            table.insert(sprites, sprite)
        end
    end

    return sprites
end

return MagneticCeiling